File: Java/medium/P_109.java
package medium;

import utils.DataStructures.ListNode;
import utils.DataStructures.TreeNode;

public class P_109 {

    public TreeNode sortedListToBST(ListNode head) {
        return toBST(head, null);
    }

    public TreeNode toBST(ListNode head, ListNode tail) {
        if (head == tail) {
            return null;
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != tail && fast.next != tail) {
            fast = fast.next.next;
            slow = slow.next;
        }

        final TreeNode root = new TreeNode(slow.val);
        root.left = toBST(head, slow);
        root.right = toBST(slow.next, tail);
        return root;
    }
}

File: cpp/leetcode/P_109.cpp
#include <bits/stdc++.h>
#define fast_io                \
  ios::sync_with_stdio(false); \
  cin.tie(nullptr);
using namespace std;

struct ListNode {
  int val;
  ListNode* next;
  ListNode(int _val, ListNode* _next) {
    val = _val;
    next = _next;
  }
};
struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int _val) { val = _val; }
};

TreeNode* sortedListToBST(ListNode* head) {
  if (head == nullptr) {
    return nullptr;
  }
  if (head->next == nullptr) {
    return new TreeNode(head->val);
  }
  ListNode* dummy = new ListNode(-1, head);
  ListNode* slow = dummy;
  ListNode* fast = dummy;
  while (fast->next != nullptr && fast->next->next != nullptr) {
    fast = fast->next->next;
    slow = slow->next;
  }
  ListNode* right = slow->next->next;
  ListNode* root = slow->next;
  slow->next->next = nullptr;
  slow->next = nullptr;
  TreeNode* res = new TreeNode(root->val);
  res->left = sortedListToBST(dummy->next);
  res->right = sortedListToBST(right);
  return res;
}
